components/nexus-repository/src/main/java/org/sonatype/nexus/repository/httpclient/FilteredHttpClient.java

```diff
@@ -0,0 +1,354 @@
+import org.apache.http.conn.ConnectionPoolTimeoutException;
+import org.apache.http.params.HttpParams;
+import org.apache.http.protocol.HttpContext;
+import org.joda.time.DateTime;
+import org.joda.time.Duration;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * @since 3.0
+ */
+public class FilteredHttpClient
+    extends ComponentSupport
+    implements HttpClient, Closeable
+{
+
+  private final HttpClient delegate;
+
+  private final boolean blocked;
+
+  private HttpHost mainTarget;
+
+  private DateTime blockedUntil;
+
+  private Thread checkThread;
+
+  private final boolean autoBlock;
+
+  private final NumberSequence autoBlockSequence;
+
+  private RemoteConnectionStatus status;
+
+  public FilteredHttpClient(final HttpClient delegate,
+                            final HttpClientConfig config)
+  {
+    this.delegate = checkNotNull(delegate);
+    blocked = checkNotNull(config).getConnectionConfig().isBlocked();
+    autoBlock = config.getConnectionConfig().shouldAutoBlock();
+    status = new RemoteConnectionStatus(blocked ? "Remote Manually Blocked" : "Unknown");
+    // TODO shall we use config.getConnectionConfig().getTimeout() * 2 as in NX2?
+    autoBlockSequence = new FibonacciNumberSequence(Time.seconds(40).toMillis());
+  }
+
+  private <T> T filter(final HttpHost target, final Filterable<T> filterable) throws IOException {
+    // main target is the first accessed target
+    if (mainTarget == null) {
+      mainTarget = target;
+    }
+    // we only filter requests to our main target
+    if (!target.equals(mainTarget)) {
+      return filterable.call();
+    }
+    if (blocked) {
+      throw new IOException("Remote Manually Blocked");
+    }
+    DateTime blockedUntilCopy = this.blockedUntil;
+    if (autoBlock && blockedUntilCopy != null && blockedUntilCopy.isAfterNow()) {
+      throw new IOException("Remote Auto Blocked");
+    }
+    try {
+      T result = filterable.call();
+      if (autoBlock) {
+        synchronized (this) {
+          if (blockedUntil != null) {
+            blockedUntil = null;
+            checkThread.interrupt();
+            checkThread = null;
+            autoBlockSequence.reset();
+          }
+        }
+      }
+      status = new RemoteConnectionStatus("Remote Available");
+      return result;
+    }
+    catch (IOException e) {
+      if (isRemoteUnavailable(e)) {
+        if (autoBlock) {
+          synchronized (this) {
+            // avoid some other thread already increased the sequence
+            if (blockedUntil == null || blockedUntil.isBeforeNow()) {
+              blockedUntil = DateTime.now().plus(autoBlockSequence.next());
+              if (checkThread != null) {
+                checkThread.interrupt();
+              }
+              String uri = target.toURI();
+              // TODO maybe find different means to schedule status checking
+              checkThread = new Thread(new CheckStatus(uri, blockedUntil), "Check Status " + uri);
```

Comment:
Discussed about this with Jason and we postponed having a thread pool till we have a better infrastructure in place. Put the TODO there so we revisit